consensus/src/pos/mod.rs

```diff
@@ -124,15 +133,13 @@ where
         .get_gen_block_index(header.prev_block_id())?
         .ok_or_else(|| ConsensusPoSError::PrevBlockIndexNotFound(header.get_id()))?;
 
-    let kernel_output = get_kernel_output(pos_data, utxos_view)?;
-
-    let epoch_index =
-        chain_config.epoch_index_from_height(&prev_block_index.block_height().next_height());
-
-    let random_seed = randomness_of_sealed_epoch(chain_config, epoch_index, block_index_handle)?;
+    let current_height = prev_block_index.block_height().next_height();
+    let random_seed = randomness_of_sealed_epoch(chain_config, current_height, block_index_handle)?;
 
+    let kernel_output = get_kernel_output(pos_data, utxos_view)?;
```

Comment:
I don't think this will work. We have to think about this better.

The problem is that we're checking this at the wrong stage. This should be checked when connecting blocks. The reason is that this depends on the history of the blockchain, and anything that depends on the history should be strictly done when connecting the block. Otherwise we'll potentially mischaracterize spends as double-spends when there are forks.

Reply:
not sure that I understand what you mean. Kernel output is required to get the vrf pub key and calculate hash to compare with target. So basically the whole PoS consensus depends on the history.

Reply:
This is the paradox I mentioned in the meeting today.

Reply:
yeah, it doesn't work already. No reorg can be performed because of this - tried to add `pos_processing_tests::check_pool_balance_after_reorg` test. I believe it's the next thing to do

Reply:
Reorgs can be done as long as there's no conflict. The bigger issue is that we won't be able to distinguish between a true double-spend attempt and an honest attempt to add a block, and an honest reorg that causes respending relative to a previous state.